### ISSUE TITLE
Add voice message transcription like iOS 17's iMessage

### ISSUE THREAD

**Issue by Avi0n** [vector-im/element-x-ios#2031]
New in iOS 17, iMessage now transcribes voice messages. This feature would be useful in Element X when you receive a voice message, and you're not in an area where you're able to listen to it. It would also be a helpful accessibility feature.

![image](https://github.com/vector-im/element-x-ios/assets/14863961/ece2ba3c-c5fe-4d34-b069-7191b2873543)

**Comment by HarHarLinks** [comment]
Is this a sender or a receiver feature (in iMessage)?
Generally, it would make sense to be a sender one so everyone gets the same transcription. This would need to be a protocol level feature then, presumably using extensible events with m.voice falling back to m.text.
In either case, this should also be an Element Meta feature, which means AI SST needs to be supported on all platforms.

**Comment by Avi0n** [comment]
In iMessage, it's a receiver feature. The audio message is transcribed on-device on the receiver's end.

**Comment by HarHarLinks** [comment]
Correct, I was assuming that Element would prioritize delivering this features consistently across its platforms, requiring support on all of them anyway. In which case having the sender do it is the better choice for the stated reason.

**Comment by d-ryabov** [comment]
Were there any updates on the task? This functionality would be really interesting.

**Comment by d-ryabov** [comment]
I’ve opened a new issue to describe how this could be implemented using the latest iOS APIs: 

https://github.com/element-hq/element-x-ios/issues/5285

That issue focuses on an iOS 26+ only approach using on-device speech recognition (`SpeechAnalyzer` / `SpeechTranscriber`), which might be a more realistic path forward for adding voice message transcription in the current architecture.

**Comment by d-ryabov** [comment]
https://github.com/element-hq/element-x-ios/issues/5285 was closed as a duplicate, so I'm transferring the information from it here:



**Is your feature request related to a problem? Please describe.** It is often inconvenient or impossible to listen to voice messages (public places, at work, accessibility needs). Being able to quickly read what was said would make Element X much easier to use in these situations. There is an older transcription request that seems stalled and does not reflect the newer on-device speech APIs available on recent iOS versions.

**Describe the solution you'd like** On iOS 26+, Apple provides `SpeechAnalyzer` / `SpeechTranscriber` APIs for high-quality on-device speech-to-text. The idea is to add an optional, iOS-only transcription UI for voice messages that does not require any extra backend and keeps audio fully on the device.

Preferred UX: a small “Transcribe” button on each voice message bubble (similar in spirit to Telegram). When tapped, it runs on-device transcription and shows the text under the message.

As an alternative, Element X could show a short preview of the first sentences under the voice message and expand to the full text on tap, in a more iOS-style way.

**Describe alternatives you've considered**

* Server-side transcription: would require infrastructure, introduce cost and privacy concerns, and does not fit well with Element’s privacy-first positioning.
* Doing nothing: keeps voice messages audio-only and less usable in many real-world contexts.

**Additional context** This would be an iOS 26+ only enhancement, guarded by availability checks for the new Speech framework APIs, so it does not affect other platforms or older iOS versions. Using the system `SpeechAnalyzer` / `SpeechTranscriber` APIs keeps the implementation relatively contained while showcasing that Element X adopts modern iOS capabilities in a privacy-preserving way.

Relevant Apple resources for implementation:

* [WWDC25 session “Bring advanced speech-to-text to your app with SpeechAnalyzer”](https://developer.apple.com/videos/play/wwdc2025/277/)
* [Speech framework overview](https://developer.apple.com/documentation/speech)
* [SpeechTranscriber documentation](https://developer.apple.com/documentation/speech/speechtranscriber)
